### PR TITLE
fix incorrect `__richcmp__` for `eq_int` only simple enums

### PR DESCRIPTION
**Pull request by Icxolu**
This fixes a regression that I accidentally introduced in #4210.  @Zyell noticed it in https://github.com/PyO3/pyo3/pull/4202#issuecomment-2142991466 🙏 

The match arms for the `eq_int` comparison would not be generated as intended if `eq` was not also specified. This should now generate the correct arms.

### CODE REVIEW

**Comment by davidhewitt** [comment]
Hmm interesting case. I sort of wonder if we don't need this and it's ok if adding `eq_int` will disable the automated value equality?

Justification: users who've added `eq_int` are already customising the equality of their enum beyone the 0.21 behaviour and while it seems unlikely that users would intentionally want `eq_int` without `eq`, if we merge this patch then they might accidentally have implicit `eq` which goes away in the future when we remove the deprecated behaviour?

**Comment by Zyell** [comment]
@davidhewitt currently adding eq_int causes invalid equality checks and it deviates from the prior behavoir.  Isn't eq_int supposed to maintain the current equality behavior?  I am somewhat confused about the use cases.  I had thought it was like the following:

- `eq` only - provides the new equality check behavior (requires `PartialEq` implementation)
- `eq' and `eq_int` - provides the new equality check with fallback to the older integer comparison that was default provided by default on simple enums
- `eq_int`  only- provides only the original integer comparison behavior
- no equality parameter - maintains old behavior with deprecation warning. (this also currently gives incorrect comparisons)

Please correct me if I mistook the structure.

**Comment by davidhewitt** [comment]
I'll recheck (and I might be remembering wrong) - I thought the 0.21 behaviour matches (functionally, but not in implementation) what we now call `#[pyclass(eq, eq_int)]`... 🤔

**Comment by Icxolu** [comment]
Nice list, this is essentially what I tried to build 😅 . No equality is equivalent to `eq_int` only, so they are both currently broken. I think the question is if we want `Enum.VariantA == Enum.VariantA` to work even if only `eq_int` (and not `eq`) is provided.

I think we probably also don't want to emit a deprecation warning if `eq_int` is missing, but `eq` was added. I think I will add that fix here as well.

**Comment by Zyell** [comment]
@davidhewitt If you check this [test](https://github.com/PyO3/pyo3/blob/a646a820fd3fc6211c52291de8b23a61f7ae7c22/tests/test_enum.rs#L56) from the 0.21 release, it will not work under the same conditions on main.  

@Icxolu I agree with the deprecation warning.  Are you also thinking eq_int would require eq to also be passed then?

**Comment by Icxolu** [comment]
I was thinking to only emit the deprecation warning if neither `eq` nor `eq_int` is present. If one or both of them were added, the user made a choice, which behavior is desired, so there is no need for the warning anymore. We could require `eq` for `eq_int`, but it's not _technically_ necessary.

**Comment by davidhewitt** [comment]
> @davidhewitt If you check this [test](https://github.com/PyO3/pyo3/blob/a646a820fd3fc6211c52291de8b23a61f7ae7c22/tests/test_enum.rs#L56) from the 0.21 release, it will not work under the same conditions on main.

Right I see, so we already broke it. Can we perhaps add that test of the 0.21 behaviour back under an `#[allow(deprecated)]` guard? That would help me be sure that this patch is doing what we want.

> I was thinking to only emit the deprecation warning if neither `eq` nor `eq_int` is present. If one or both of them were added, the user made a choice, which behavior is desired, so there is no need for the warning anymore. We could require `eq` for `eq_int`, but it's not _technically_ necessary.

Fully agree with this 👍

**Comment by Icxolu** [comment]
> Can we perhaps add that test of the 0.21 behaviour back under an `#[allow(deprecated)]` guard?

That's a good idea. I added a `deprecated` module with the relevant tests.

File: tests/test_enum.rs [hunk 1]
```diff
@@ -280,3 +318,59 @@ fn test_complex_enum_with_hash() {
...
+    #[test]
+    fn test_custom_discriminant() {
+        Python::with_gil(|py| {
+            #[allow(non_snake_case)]
+            let CustomDiscriminant = py.get_type_bound::<CustomDiscriminant>();
+            let one = Py::new(py, CustomDiscriminant::One).unwrap();
+            let two = Py::new(py, CustomDiscriminant::Two).unwrap();
+            py_run!(py, CustomDiscriminant one two, r#"
+            assert CustomDiscriminant.One == one
+            assert CustomDiscriminant.Two == two
```
Comment (davidhewitt) [inline tests/test_enum.rs:371]:
```suggestion
            assert CustomDiscriminant.One == one
            assert CustomDiscriminant.Two == two
            assert CustomDiscriminant.One == 1
            assert CustomDiscriminant.Two == 2
```

File: tests/test_enum.rs [hunk 2]
```diff
@@ -280,3 +318,59 @@ fn test_complex_enum_with_hash() {
...
+            assert one != two
```
Comment (davidhewitt) [inline tests/test_enum.rs:372]:
```suggestion
            assert one != two
            assert CustomDiscriminant.One != 2
            assert CustomDiscriminant.Two != 1
```

File: tests/test_enum.rs [hunk 3]
```diff
@@ -221,6 +221,44 @@ fn test_renaming_all_enum_variants() {
...
+}
+
+#[test]
+fn test_simple_enum_eq_int_only() {
+    Python::with_gil(|py| {
+        let var1 = Py::new(py, EqIntOnly::VariantA).unwrap();
+        let var2 = Py::new(py, EqIntOnly::VariantA).unwrap();
+        let var3 = Py::new(py, EqIntOnly::VariantB).unwrap();
+        py_assert!(py, var1 var2, "var1 == var2");
+        py_assert!(py, var1 var3, "var1 != var3");
```
Comment (davidhewitt) [inline tests/test_enum.rs:238]: Does this imply that by having `eq_int` that normal equality "works" anyway (but via integer conversion presumably)?

If so, maybe it makes sense to have `eq_int` require `eq` to make it not a surprise that value equality works? We can always relax that requirement later I guess.

**Review by davidhewitt — COMMENTED** [review]
Thanks, the new test cases really help! I think we're nearly there...

File: tests/test_enum.rs (hunk 3 above)
Comment (Icxolu) [reply tests/test_enum.rs:238]: > Does this imply that by having `eq_int` that normal equality "works" anyway (but via integer conversion presumably)?

Yes, this is correct. 

> If so, maybe it makes sense to have `eq_int` require `eq` to make it not a surprise that value equality works? We can always relax that requirement later I guess.

Sure, I would love to have value comparisons only work with `eq`, but I did not see a good way to archive that.

**Review by davidhewitt — APPROVED** [review]
LGTM, thanks! 💯

**Pull request merged**